coverage.c

```diff
@@ -0,0 +1,590 @@
+
+    data = (bam_aux_t **)calloc(n_bam_files, sizeof(bam_aux_t*)); // data[i] for the i-th BAM file
+
+    for (i = 0; i < n_bam_files; ++i) {
+        int rf;
+        data[i] = (bam_aux_t *) calloc(1, sizeof(bam_aux_t));
+        data[i]->fp = sam_open_format(argv[optind+i], "r", &ga.in); // open BAM
+
+        if (data[i]->fp == NULL) {
+            print_error_errno("coverage", "Could not open \"%s\"", argv[optind+i]);
+            status = EXIT_FAILURE;
+            goto coverage_end;
+        }
+        rf = SAM_FLAG | SAM_RNAME | SAM_POS | SAM_MAPQ | SAM_CIGAR | SAM_SEQ;
+        if (opt_min_baseQ) rf |= SAM_QUAL;
+
+        // Set CRAM options on file handle - returns 0 on success
+        if (hts_set_opt(data[i]->fp, CRAM_OPT_REQUIRED_FIELDS, rf)) {
+            print_error_errno("coverage", "Failed to set CRAM_OPT_REQUIRED_FIELDS value");
+            status = EXIT_FAILURE;
+            goto coverage_end;
+        }
+        if (hts_set_opt(data[i]->fp, CRAM_OPT_DECODE_MD, 0)) {
+            print_error_errno("coverage", "Failed to set CRAM_OPT_DECODE_MD value");
+            status = EXIT_FAILURE;
+            goto coverage_end;
+        }
+        data[i]->min_mapQ = opt_min_mapQ;            // set the mapQ filter
+        data[i]->min_len  = opt_min_len;             // set the qlen filter
+        data[i]->hdr = sam_hdr_read(data[i]->fp);    // read the BAM header
+        if (data[i]->hdr == NULL) {
+            print_error_errno("coverage", "Couldn't read header for \"%s\"", argv[optind+i]);
+            status = EXIT_FAILURE;
+            goto coverage_end;
+        }
+
+        // Lookup region if specified
+        if (opt_reg) { // if a region is specified
+            hts_idx_t *idx = sam_index_load(data[i]->fp, argv[optind+i]);  // load the index
+            if (idx == NULL) {
+                print_error_errno("coverage", "can't load index for \"%s\"", argv[optind+i]);
+                status = EXIT_FAILURE;
+                goto coverage_end;
+            }
+            data[i]->iter = sam_itr_querys(idx, data[i]->hdr, opt_reg); // set the iterator
+            hts_idx_destroy(idx); // the index is not needed any more; free the memory
+            if (data[i]->iter == NULL) {
+                print_error_errno("coverage", "can't parse region \"%s\"", opt_reg);
+                status = EXIT_FAILURE;
+                goto coverage_end;
+            }
+        }
+    }
+
+    if (opt_print_tabular && opt_print_header)
+        fputs("#CHROM\tSTART\tEND\tN_READS\tN_COVERED_BASES\tPERCENT_COVERED\tAVG_COV\tAVG_BASEQ\tAVG_MAPQ\n", file_out);
+
+
+    h = data[0]->hdr; // easy access to the header of the 1st BAM
+    covered_tids = calloc(h->n_targets, sizeof(bool));
+    stats_aux_t *stats = calloc(1, sizeof(stats_aux_t));
+
+    bool reg_sets_end = false;
+    if (opt_reg) {
+        stats->beg = data[0]->iter->beg; // and to the parsed region coordinates
+        stats->end = data[0]->iter->end;
```

Comment:
This works only provided we fully specify the reference.  Eg 20:1000000-20000000 will show 10M to 20M on the X-axis.  However if we just do "20" then it'll go from 1 to INT_MAX (2.147Gb).  It should check for these cases and lookup in the header to find the sequence extents.  There is a `target_len` array in bam_hdr_t for this (indexed on tid).